[WIP] add support for api key source on api gateway

pablo-ruth opened the pull request:
Fixes #4425

Changes proposed in this pull request:

* Add support for API Key source in API Gateway

pablo-ruth commented:
Acc Checks OK

```
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccAWSAPIGatewayRestApi -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAWSAPIGatewayRestApi_basic
--- PASS: TestAccAWSAPIGatewayRestApi_basic (75.50s)
=== RUN   TestAccAWSAPIGatewayRestApi_EndpointConfiguration
--- PASS: TestAccAWSAPIGatewayRestApi_EndpointConfiguration (86.75s)
=== RUN   TestAccAWSAPIGatewayRestApi_policy
--- PASS: TestAccAWSAPIGatewayRestApi_policy (75.17s)
=== RUN   TestAccAWSAPIGatewayRestApi_openapi
--- PASS: TestAccAWSAPIGatewayRestApi_openapi (58.33s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       295.777s
```

bflad commented:
Hi @pablo-ruth 👋 Thanks so much for submitting this! I think you're right on track so far -- could you please take a look at:

* Implementing the patch operation to update the value in `resourceAwsApiGatewayRestApiUpdateOperations`? Currently if you attempted to change the value of this attribute for an existing resource in Terraform it will continually show as needing to be updated but not be able do anything. The API Gateway API seems to support the `replace` operation for `/apiKeySource` according to https://docs.aws.amazon.com/apigateway/api-reference/link-relation/restapi-update/
* Adding an acceptance test check for the value of `api_key_source` in `TestAccAWSAPIGatewayRestApi_basic` -- something like this should do it: `resource.TestCheckResourceAttr("aws_api_gateway_rest_api.test", "api_key_source", "HEADER"),`
* Implementing a separate test case and configuration that tries setting the value to `AUTHORIZER` and updating it back to `HEADER`? Feel free to copy-paste something similar to the `policy` test and config should get you really close.

Please let us know if you do not have time to implement these or have any questions - thanks again!

pablo-ruth commented:
Hi @bflad,

Thanks for your guidance. I implemented patch method, added acc test and doc. Acc tests pass.

```
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccAWSAPIGatewayRestApi -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAWSAPIGatewayRestApi_basic
--- PASS: TestAccAWSAPIGatewayRestApi_basic (62.29s)
=== RUN   TestAccAWSAPIGatewayRestApi_EndpointConfiguration
--- PASS: TestAccAWSAPIGatewayRestApi_EndpointConfiguration (68.81s)
=== RUN   TestAccAWSAPIGatewayRestApi_api_key_source
--- PASS: TestAccAWSAPIGatewayRestApi_api_key_source (65.33s)
=== RUN   TestAccAWSAPIGatewayRestApi_policy
--- PASS: TestAccAWSAPIGatewayRestApi_policy (57.70s)
=== RUN   TestAccAWSAPIGatewayRestApi_openapi
--- PASS: TestAccAWSAPIGatewayRestApi_openapi (46.05s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       300.187s
```

bflad commented:
This has been released in [version 1.22.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1220-june-05-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
